packages/common/src/ArcGISContextManager.ts

```diff
@@ -0,0 +1,242 @@
+   * instantiate an instance
+   * @param opts
+   */
+  private constructor(opts: IArcGISContextManagerOptions) {
+    // Having a unique id makes debugging easier
+    this.id = new Date().getTime();
+    if (opts.debug) {
+      this._debug = opts.debug;
+    }
+    this.log(`ArcGISContextManager:ctor: Creating ${this.id}`);
+
+    if (opts.authentication) {
+      this._authentication = opts.authentication;
+      this._portalUrl = this._authentication.portal.replace(
+        "/sharing/rest",
+        ""
+      );
+      this._hubUrl = getHubApiFromPortalUrl(this._portalUrl);
+    } else if (opts.portalUrl) {
+      this._portalUrl = opts.portalUrl;
+      this._hubUrl = getHubApiFromPortalUrl(this._portalUrl);
+    } else {
+      this._hubUrl = getHubApiFromPortalUrl(this._portalUrl);
+    }
+
+    if (opts.portal) {
+      this._portalSelf = cloneObject(opts.portal);
+    }
+    if (opts.currentUser) {
+      this._currentUser = cloneObject(opts.currentUser);
+    }
+  }
+
+  /**
+   * Static async Factory
+   * @param opts
+   * @returns
+   */
+  public static async create(
+    opts: IArcGISContextManagerOptions = {}
+  ): Promise<ArcGISContextManager> {
+    const ctx = new ArcGISContextManager(opts);
+    await ctx.initialize();
+    return ctx;
+  }
+
+  /**
+   * Set the Authentication (UserSession) for the context.
+   * This should be called when a user signs into a running
+   * application.
+   * @param auth
+   */
+  async setAuthentication(auth: UserSession): Promise<void> {
+    this._authentication = auth;
+    this._portalUrl = auth.portal.replace("/sharing/rest", "");
+    await this.initialize();
+  }
+
+  /**
+   * Clear the Authentication (UserSession). This should be
+   * called when a user signs out of an application, but
+   * the application continues running
+   */
+  clearAuthentication(): void {
+    // Reset the portalUrl from the org url to the base url
+    // for ArcGIS Enterprise, we just leave the _portalUrl as-is
+    if (!this._context.isPortal) {
+      this._portalUrl = getPortalBaseFromOrgUrl(this._portalUrl);
+    }
+    // Clear the auth, portalSelf and currentUser props
+    this._authentication = null;
+    this._portalSelf = null;
+    this._currentUser = null;
+    this._context = new ArcGISContext({
+      id: this.id,
+      portalUrl: this._portalUrl,
+      hubUrl: this._hubUrl,
+    });
+  }
+
+  /**
+   * Return a reference to the current state.
+   * When `.setAuthentication()` or `.clearAuthenentication()` are
+   * called, the state will be re-created. This is done so frameworks
+   * like React or Ember can detect changes.
+   */
+  get context(): IArcGISContext {
+    return this._context;
+  }
+
+  /**
+   * @internal
+   * Log debugging messages to the console
+   * @param message
+   */
+  private log(message: string): void {
+    if (this._debug) {
+      // tslint:disable-next-line:no-console
+      console.info(message);
```

Comment:
not a blocker for this PR, but probably should use https://esri.github.io/hub.js/api/common/Logger